module/zfs/zio.c

```diff
@@ -3028,21 +3112,27 @@ zio_done(zio_t *zio)
 			zio_cksum_report_t *zcr = zio->io_cksum_report;
 			uint64_t align = zcr->zcr_align;
 			uint64_t asize = P2ROUNDUP(zio->io_size, align);
-			char *abuf = zio->io_data;
+			char *abuf;
+			abd_t *adata = zio->io_data;
 
 			if (asize != zio->io_size) {
-				abuf = zio_buf_alloc(asize);
-				bcopy(zio->io_data, abuf, zio->io_size);
-				bzero(abuf+zio->io_size, asize-zio->io_size);
+				adata = abd_alloc_linear(asize);
+				abd_copy(adata, zio->io_data, zio->io_size);
+				abd_zero_off(adata, asize-zio->io_size,
+				    zio->io_size);
 			}
 
+			abuf = abd_borrow_buf_copy(adata, asize);
```

Comment:
adata may be NULL (e.g. if it's a zio_free).  in this case you need to set abuf to NULL; abd_borrow_buf_copy(NULL) won't work.